awx/main/consumers.py

```diff
@@ -12,12 +13,53 @@
 from channels.generic.websocket import AsyncJsonWebsocketConsumer
 from channels.layers import get_channel_layer
 from channels.db import database_sync_to_async
+from channels_redis.core import RedisChannelLayer
 
 
 logger = logging.getLogger('awx.main.consumers')
 XRF_KEY = '_auth_user_xrf'
 
 
+class ExpiringCache(collections.defaultdict):
+    def __init__(self, default, ttl=60, *args, **kw):
+        collections.defaultdict.__init__(self, default)
+        self._expires = collections.OrderedDict()
+        self.ttl = ttl
+
+    def __setitem__(self, k, v):
+        collections.defaultdict.__setitem__(self, k, v)
+        self._expires[k] = time.time() + self.ttl
+
+    def __delitem__(self, k):
+        try:
+            collections.defaultdict.__delitem__(self, k)
+        except KeyError:
+            # RedisChannelLayer itself _does_ periodically clean up this
+            # dictionary (e.g., when exceptions like asyncio.CancelledError
+            # occur)
+            pass
+
+    def expire(self):
+        expired = []
+        for k in self._expires.keys():
+            if self._expires[k] < time.time():
+                expired.append(k)
+            else:
+                # as this is an OrderedDict, every key after this
+                # was inserted *later*, so if _this_ key is *not* expired,
+                # the ones after it aren't either (so we can stop iterating)
+                break
+        for k in expired:
+            del self._expires[k]
+            del self[k]
```

Comment:
would `self.pop(k, None)` work the same w/out the `del` + `__delitem__` override?

Reply:
Yes, **but** the reason I defined a `__delitem__` is because the redis cache *itself* is asyncio code that sometimes *also* deletes these keys as part of callbacks.  So there are situations where there's a race to see which coroutine will clean up first.  So I decided to just make KeyErrors on `del` just be a <shrug> to simplify my life.

Reply:
Ah. Yep. I see now: https://github.com/django/channels_redis/blob/master/channels_redis/core.py#L462

Unless you override the `del`, `channels_redis` can still hit the error if it loses the race.